### PR TITLE
Let tests run when Python executable path contains whitespace

### PR DESCRIPTION
**Pull request by McSinyx**
I hope that this solves #3369.  Speaking from experience (of maintaining widgets fetching data from command-line programs), quoting shell is a huge pain, that's why I'm favoring imputing commands via lists.

If any change is too excessive, feel free to tell me to undo it.

### CODE REVIEW

File: runtests.py [hunk 1]
```diff
@@ -1713,9 +1713,9 @@ def __init__(self, treefile, workdir, cleanup_workdir=True, stats=None,
         self.cleanup_workdir = cleanup_workdir
         self.stats = stats
         self.capture = capture
-        cython_syspath = [self.cython_root]
+        cython_syspath = [EndToEndTest.cython_root]
```
Comment (scoder) [inline runtests.py]: Why this change?

File: runtests.py [hunk 2]
```diff
@@ -1766,21 +1762,19 @@ def runTest(self):
...
+                                 'etoe-build' if 'setup.py' in command else 'etoe-run'):
                 if self.capture:
                     p = subprocess.Popen(command,
-                                     stderr=subprocess.PIPE,
-                                     stdout=subprocess.PIPE,
-                                     shell=True,
-                                     env=env)
+                                         stderr=subprocess.PIPE,
+                                         stdout=subprocess.PIPE,
+                                         shell=True,
```
Comment (scoder) [inline runtests.py]: `command` is a list now, right? Shouldn't this option be removed then? (Same below)

File: Cython/TestUtils.py [hunk 3]
```diff
@@ -184,34 +186,35 @@ def visit_CompilerDirectivesNode(self, node):
...
+                        command = shlex.split(line)
+                        if not command: continue
+                        # In Python 3: prog, *args = command
+                        prog, args = command[0], command[1:]
+                        try:
+                            header.append(programs[prog]+args)
+                        except KeyError:
+                            header.append(command)
+        finally:
+            if not getattr(cur_file, 'closed', True): cur_file.close()
```
Comment (scoder) [inline Cython/TestUtils.py]: Separate lines.

File: Cython/TestUtils.py [hunk 4]
```diff
@@ -184,34 +186,35 @@ def visit_CompilerDirectivesNode(self, node):
...
+        try:
+            for line in f:
+                if line.startswith('#####'):
+                    filename = line.strip().strip('#').strip().replace('/', os.path.sep)
+                    path = os.path.join(workdir, filename)
+                    if not os.path.exists(os.path.dirname(path)):
+                        os.makedirs(os.path.dirname(path))
+                    cur_file.close()
+                    cur_file = open(path, 'w')
+                elif not getattr(cur_file, 'closed', True):
```
Comment (scoder) [inline Cython/TestUtils.py]: Why not keep the much simpler `is None`? (Same below) This change makes the unnecessary assumption that whatever `open()` returns has a `closed` attribute that changes with a call to `.close()`. I don't see a need for that assumption.

File: runtests.py (hunk 1 above)
Comment (McSinyx) [reply runtests.py]: This is to clarify `cython_root` is class attribute.  I think it'd fit since this commit touches the attribute, but I can undo it if it creates blame noise.

Comment (scoder) [reply runtests.py]: `self.xyz` is just fine for attributes defined in the class, but additionally works for subclasses and instances. I do not see a need to make this code less generic than it is.

File: Cython/TestUtils.py (hunk 4 above)
Comment (McSinyx) [reply Cython/TestUtils.py]: > Why not keep the much simpler `is None`?

You're right, for this case it's only to guard against the case when no output file has ever been opened.

> Same below

In the `finally` clause, we're guarding against two cases:
* No output file has ever been opened: checking for None works fine
* `cur_file` is assigned to a currently opening file by the first `if`: if the statement is reached because an exception is raised, it should try to close the file.  I'm not in favor of a `try` that big, but I'm not capable of making it any more explicit.

> This change makes the unnecessary assumption that whatever `open()` returns has a `closed` attribute that changes with a call to `.close()`.

No assumption is made, [since what ever `open` returns is based on `IOBase`, which define the attribute `closed`](https://docs.python.org/2/library/io.html#io.IOBase.closed).

File: runtests.py (hunk 2 above)
Comment (McSinyx) [reply runtests.py]: I think so.  Let me try if it works.

File: Cython/TestUtils.py (hunk 4 above)
Comment (scoder) [reply Cython/TestUtils.py]: There used to be this code, which maps all cases back to `None`, and which I think is fairly explicit:
```python
               if cur_file is not None:
                    f, cur_file = cur_file, None
                    f.close()
```

File: runtests.py (hunk 1 above)
Comment (McSinyx) [reply runtests.py]: I understand, I'll revert to the previous code.  I made the change after unable to search for `self.cython_root` anywhere and thought a search to `EndToEndTest` would be faster to locate the variable's value.

Comment (scoder) [reply runtests.py]: No need, I already updated your branch.

File: Cython/TestUtils.py (hunk 4 above)
Comment (McSinyx) [reply Cython/TestUtils.py]: Although I don't like the way this is handled, I think I'm OK with it if it's safer this way.

Comment (scoder) [reply Cython/TestUtils.py]: It's ok to disagree on solutions. Here's why I chose the original code over your proposal.

IMHO (!), `getattr()` has two use cases. The good one is to read an attribute by string name. And the smelly one is to check for an attribute and read it if it's there. It feels smelly because it's fairly broad – there may be many reasons why such an attribute does not exist, and some of them may be bugs or at least unintended. And for the innocent reader, it's difficult to understand what the originally considered and correctly covered reasons were.

In your case, I assume that you were only expecting `None` to hit the "no such attribute" case. That may be so – as long as things really work as expected and no-one replaces the builtin `open`, for example (fine, that's probably acceptable). For a future reader, however, it may take a while to understand that this is the only intended case, and that changing this code will not lead to unintended bugs due to other, not so obvious cases.

If, instead, I write `is not None`, then it's immediately clear that this is the only intended case. And looking at the other usages will somewhat quickly make it clear in which cases `None` can appear. For me, that's a reason to prefer the original version.

**Comment by scoder** [comment]
Thanks, @McSinyx !

File: Cython/TestUtils.py (hunk 4 above)
Comment (McSinyx) [reply Cython/TestUtils.py]: Thanks for the insight, especially the forward-compatibility/gotchas!
